sentences/de/_common.yaml

```
@@ -288,6 +312,20 @@ expansion_rules:
   sind_wurden: "(sind|wurden)"
   position: "{position} [Prozent|%]"
   volume: "{volume_level} [Prozent|%]"
+
+  # Timers
+  timer_set: "(starte|setze|stelle|erstelle)"
+  timer_cancel: "(beende|stoppe)"
+  timer_duration_seconds: "{timer_seconds:seconds} Sekunde[n]"
+  timer_duration_minutes: "{timer_minutes:minutes} Minute[n][ [und ]{timer_seconds:seconds} Sekunde[n]]"
+  timer_duration_hours: "{timer_hours:hours} Stunde[n][ [und ]{timer_minutes:minutes} Minute[n]][ [und ]{timer_seconds:seconds} Sekunde[n]]"
+  timer_duration: "<timer_duration_seconds>|<timer_duration_minutes>|<timer_duration_hours>"
```

```suggestion
  timer_duration_seconds: "{timer_seconds:seconds} Sekunde[n]"
  timer_duration_minutes: "{timer_minutes:minutes} Minute[n]["
  timer_duration_hours: "{timer_hours:hours} Stunde[n]"
  timer_duration: "[<timer_duration_hours>] [und] [<timer_duration_minutes>] [und] [<timer_duration_seconds>]"
```

i would suggest something like this (_it is not tested, but the idea should be clear_) to avoid repeating all previous ones (_this counts also for the following `timer_start_` expansion rules_)

---

I followed @synesthesiam's approach that is also used in the current implementations of timers in RO, RU, NL. I do think your approach is better but I am not sure how Mike wants to progress with this and I think this should rather be discussed globally so it can be changed in all languages and not just DE.

---

this is nothing "global", but specific to each language.

---

Using that approach I can not get a sentence like `Stelle Timer namens Pizza für 5 Minuten` to match correctly. The `timer_name` slot is matched in a greedy fashion as `pizza für 5 minuten` instead of `pizza` as with the approach you suggested, `timer_duration` can be completely empty (everything in the expansion is optional). I assume that's why Mike chose to go the rather strange route so that the `timer_duration` slot must always be non-empty.

**Expansions**
```
...
  timer_duration_seconds: "{timer_seconds:seconds} Sekunde[n]"
  timer_duration_minutes: "{timer_minutes:minutes} Minute[n]"
  timer_duration_hours: "{timer_hours:hours} Stunde[n]"
  timer_duration: "[<timer_duration_hours>] [und] [<timer_duration_minutes>] [und] [<timer_duration_seconds>]"
...
```

**Sentences**
```
...
- "<timer_set>[ einen] Timer namens {timer_name:name} für <timer_duration>"
...
```

---

mehhh ... the cause is that in `timer_duration: "[<timer_duration_hours>] [und] [<timer_duration_minutes>] [und] [<timer_duration_seconds>]` everything is optional, so even `""` will match this expansion rule, so the sentences in thew following order will match at first for `Stelle Timer namens Pizza für 5 Minuten`

```yaml
          - "<timer_set> [einen] <timer_duration> Timer namens {timer_name:name}"
          - "<timer_set> [einen] Timer namens {timer_name:name} für <timer_duration>"
```

when switching the order, than it works

```yaml
          - "<timer_set> [einen] <timer_duration> Timer namens {timer_name:name}"
          - "<timer_set> [einen] Timer namens {timer_name:name} für <timer_duration>"
```

```
$ python3 -m script.intentfest parse --language de --context area Wohnzimmer --sentence 'Stelle Timer namens Pizza für 5 Minuten'
{
  "text": "Stelle Timer namens Pizza für 5 Minuten",
  "match": true,
  "intent": "HassStartTimer",
  "slots": {
    "name": "pizza ",
    "minutes": 5
  },
  "context": {
    "area": "Wohnzimmer"
  },
  "response_key": "default",
  "response": "Timer gestartet"
}
```


but we should not have to take care about order of sentences (_they should not cover each other_), so we have to go with your approach